OpenChange/MAPIStoreMailMessage.m

```diff
@@ -396,36 +412,99 @@ - (void) _fetchBodyData
             encoding = @"7-bit";
 
           /* We should provide a case for each of the types in acceptedMimeTypes */
-          if ([mimeType isEqualToString: @"text/html"] && !mailIsEvent)
-            {
-              content = [content bodyDataFromEncoding: encoding];
-              [htmlContent appendData: content];
-            }
-          else if ([mimeType isEqualToString: @"text/plain"] && !mailIsEvent)
+          if (!mailIsEvent)
             {
+              NSString *charset;
+              NSStringEncoding stringEncoding;
+              NSString *stringValue; 
+              BOOL html;
+              if ([mimeType isEqualToString: @"text/html"])
+                {
+                  html = YES;
+                  
+                }
+              else if ([mimeType isEqualToString: @"text/plain"])
+                {
+                  html = NO;
+                }
+              else
+                {
+                  [self warnWithFormat: @"Unsupported MIME type for non-event body part: %@.",
+                        mimeType];
+                  continue;
+                }
+                          
               content = [content bodyDataFromEncoding: encoding];
-              NSString *charset = [bodyPartsCharsets objectForKey: key];
+              charset = [bodyPartsCharsets objectForKey: key];
+              stringEncoding = [NSString stringEncodingForEncodingNamed: headerCharset];
+              stringValue = nil;
               if (charset)
                 {
-                  NSString *stringValue = [content bodyStringFromCharset: charset];
-                  [textContent appendData: [stringValue dataUsingEncoding: NSUTF8StringEncoding]];
+                  if ([charset isEqualToString: headerCharset] || !stringEncoding)
```

Comment:
this comparative is wrong because you can have `utf-8`, `utf8`, `UTF-8`, `latin1`, `ISO8859-1`, `ISO-8859-1` and so on. You have to compare them as `NSStringEncoding`